update_index.py

```diff
@@ -17,20 +17,74 @@
 """
 import json
 import os
+import re
 import sys
+import time
 from distutils.version import LooseVersion
+from xmlrpc.client import Fault
 from xmlrpc.client import ServerProxy
 
 INDEX_FILE_NAME = os.path.join(os.path.dirname(__file__), "index.json")
 
 BLACKLIST = {"pytest-nbsmoke"}
 
+class RateLimitedServerProxy:
+    def __init__(self, uri):
+        self._server_proxy = ServerProxy(uri)
+
+    def browse(self, classifiers):
+        return self._rate_limit_request(self._server_proxy.browse, [classifiers])
+
+    def list_packages(self):
+        return self._rate_limit_request(self._server_proxy.list_packages)
+
+    def package_releases(self, package_name):
+        return self._rate_limit_request(self._server_proxy.package_releases, [package_name])
+
+    def release_data(self, name, version):
+        return self._rate_limit_request(self._server_proxy.release_data, [name, version])
+
+    def _rate_limit_request(self, request_method, args=None):
+        return_value = None
+
+        fetched_releases = False
+        while not fetched_releases:
+            try:
+                if args is not None:
```

Comment:
I agree, this is much cleaner. Let me do this instead

Reply:
Regarding your first comment, did you mean `args=[]` instead of `args=()`? I'm passing in a list always. I don't want to pass in a set on the off chance entries are duplicates

Reply:
In either case, I would prefer we set the default to `None` even if it's a bit clunkier. Using a default set or list could cause unwanted side effects. [This post](https://nikos7am.com/posts/mutable-default-arguments/) does a good enough job of explaining it. The [python docs](https://nikos7am.com/posts/mutable-default-arguments/) also suggest the `None` approach.

Granted we don't actually do anything to the list in this function, so the empty list/set is fine I still try to avoid it for the reasons in the above links as a general rule.

Are you opposed to keeping it `None`, @nicoddemus ?

Reply:
> Using a default set or list could cause unwanted side effects. This post does a good enough job of explaining it. The python docs also suggest the None approach.

It's definitely correct that using a mutable object like a `list` or a `set` as default value is problematic, but `args=()` is a `tuple` not a `set`. 😉 

Having said that I actually missed the really Pythonic solution, which is to use `*args` instead of `args=()`:

```python
    def release_data(self, name, version):
        return self._rate_limit_request(self._server_proxy.release_data, name, version)

    def _rate_limit_request(self, request_method, *args):
        while True:
            try:
                return request_method(*args)                
            except Fault as fault:      
                ...
```

This just let's the caller pass any arguments directly instead of packing them in a list. 👍

Reply:
> It's definitely correct that using a mutable object like a `list` or a `set` as default value is problematic, but `args=()` is a `tuple` not a `set`. 😉

Oops, confused the two for a second, sorry!